### PR TITLE
Add initial workshop project click implementation

### PR DESCRIPTION
**Pull request by MaxWofford**

### CODE REVIEW

File: app/controllers/v1/workshop_project_clicks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+      project = WorkshopProject.find(params[:project_id])
+
+      click = WorkshopProjectClick.create(
+        type_of: params[:type_of],
+        workshop_project: project,
+        user: current_user,
+        ip_address: request.ip,
+        user_agent: request.user_agent
+      )
```
Comment (zachlatta) [inline app/controllers/v1/workshop_project_clicks_controller.rb]: Handle error case. This fails silently right now.

Use `.create!` to throw an exception or to if/else.

File: app/models/workshop_project_click.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+class WorkshopProjectClick < ApplicationRecord
+  belongs_to :workshop_project
+  belongs_to :user
```
Comment (zachlatta) [inline app/models/workshop_project_click.rb]: Add a has_many to Users. We don't want unlinked associations..

File: app/models/workshop_project_click.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+
+  enum type_of: %i[live code]
```
Comment (zachlatta) [inline app/models/workshop_project_click.rb]: Should presence of `type_of` be validated?

File: spec/factories/workshop_project_clicks.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+FactoryBot.define do
+  factory :workshop_project_click do
+    workshop_project :workshop_project
+    type_of :type_of
+    ip_address { Faker::Internet.ip_v1_address }
+  end
```
Comment (zachlatta) [inline spec/factories/workshop_project_clicks.rb]: What about user?

File: spec/models/workshop_project_click_spec.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+require 'rails_helper'
+
+RSpec.describe 'V1::WorkshopProjectClick', type: :request do
+  describe 'GET /v1/projects/:id/redirect/:type_of' do
+    let(:project) { create(:workshop_project) }
+
+    # override in subtests
+    let(:headers) { {} }
+
+    before { get "/v1/projects/#{project.id}/redirect/live", headers: headers }
```
Comment (zachlatta) [inline spec/models/workshop_project_click_spec.rb]: Test both `live`, `code`, and invalid cases. Invalid cases should fail gracefully.

File: spec/models/workshop_project_click_spec.rb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+      expect(response.headers['Location']).to eq(project.live_url)
+    end
+
+    it 'creates a WorkshopProjectClick' do
+      click = WorkshopProjectClick.last
+
+      # idk
+      expect(click.workshop_project).to eq(project)
+      expect(click.ip_address).to eq('127.0.0.1')
+      expect(click.type_of).to eq('live')
```
Comment (zachlatta) [inline spec/models/workshop_project_click_spec.rb]: Yeah, this looks good. No need for the `# idk` comment. Was there something specific here I can help with?

File: app/serializers/workshop_project_click_serializer.rb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class WorkshopProjectClickSerializer < ActiveModel::Serializer
+  attributes :id, :ip_address, :referrer, :user_agent, :type_of
+  has_one :workshop_project
+  has_one :user
+end
```
Comment (zachlatta) [inline app/serializers/workshop_project_click_serializer.rb]: Why are you creating a serializer for WorkshopProjectClicks if they're never rendered?

**Comment by zachlatta** [comment]
Oh, and on the above comment, do you intend to render the click count anywhere?

**Comment by MaxWofford** [comment]
@zachlatta Just addressed each of the comments with the new changes

**Comment by MaxWofford** [comment]
re: click count

Will add it once I talk to @lachlanjc & @polytroper about where they want the info displayed

**Comment by MaxWofford** [comment]
@lachlanjc: I figure it makes sense to add 'total click count' as a field on workshop projects

File: app/serializers/workshop_project_click_serializer.rb (hunk 7 above)
Comment (MaxWofford) [reply app/serializers/workshop_project_click_serializer.rb]: This was removed in b9c6e368a4e0964d1abf86fe3aba70851497621e

**Comment by MaxWofford** [comment]
@zachlatta sure you're looking at latest? the open comments were all outdated

**Comment by lachlanjc** [comment]
@zachlatta Can you review this please?

**Pull request merged**